GVFS/GVFS.Common/Tracing/JsonTracer.cs

```diff
@@ -288,6 +300,35 @@ public void WriteStartEvent(EventMetadata metadata, Keywords keywords)
             this.WriteEvent(this.activityName, this.startStopLevel, keywords, metadata, EventOpcode.Start);
         }
 
+        void IEventListenerEventSink.OnListenerRecovery(EventListener listener)
+        {
+            // Check ContainsKey first (rather than always calling TryRemove) because ContainsKey
+            // is lock-free and recoveredListener should rarely be in failedListeners
+            if (!this.failedListeners.ContainsKey(listener))

```

Comment:
One last comment - I spent some time trying to understand whether this code could be called on multiple threads at the same time, or how it would interact with logic running asynchronously elsewhere. I think it can be, but I didn't see anythink that could obviously go wrong (i.e. - could we add a listener to the failed list, but the listener has already recovered so that it would stay on the failed list....)

Anyway - I think this all works, but I don't know if a comment around how we expect (or why we expect) that this is correct would be helpful.

Reply:
> i.e. - could we add a listener to the failed list, but the listener has already recovered so that it would stay on the failed list....

Correct, and on the next successful message from that listener we'd then detect and log recovery (just one message later). Unless we have listeners that are always going up/down/up/down between each message this will be a non-issue (and even then the issue would only be we didn't always log the failures and/or recoveries).